### DISCUSSION TITLE
Adding noise direcly to dataset to generate a dp-protected dataset

### DISCUSSION THREAD

**Discussion by santidediego** [opendp/opendp#1653] [OpenDP Library] [answered]
Reading to the library docs I understood that I can apply _measurements_ over _transformations_, for example to calculate a DP sum on a set of salaries.

However, I´d like to know if there are any functions that allow me to, given a clear dataset _d_, containing sensitive salaries data, generate a dataset _d´_ containing dp-protected salaries data, fulfilling that _sum(d)=sum(d´)_

Is this ever achievable by the library? I am starting with DP and studying its possibilities.

**Comment by mccalluc** [comment, marked as answer]
Not really, but I think it's a definitional issue, rather than a library limitation. First, it will help to clarify your unit of privacy: Here, you're probably interested in protecting the privacy of individuals, but you could imagine being concerned either with smaller units (perhaps individual paychecks) or larger (households). With that decided, let's look at one definition of DP:
> An algorithm is differentially private if by looking at the output, you cannot tell whether data from any individual [or other unit of privacy] was included in the original dataset or not.

One thing to notice is that the DP output is not the data for individuals [or other units of privacy], but an aggregation across individuals, and that's at odds with what you describe.

But could an algorithm like yours be DP? No: Let's zoom out and think about how the proposed algorithm would work at the level of the dataset:

```
your_algorithm(dataset) = sum(dataset)
your_algorithm(dataset - {my_salary}) = sum(dataset) - my_salary
```
Since by looking at the output we can see whether `my_salary` was included or not, this algorithm is not DP.

So what can we do instead? It depends on what you wanted to do with the data. Maybe a DP histogram would serve your needs: there would be a trade-off between the size of the buckets, and the accuracy of the results. Or maybe you want a set of numbers that you can use for non-DP statistics instead: In that case you might look at the literature around synthetic datasets (which could be built from histograms, or with other techniques).

Hope that helps.

**Reply by mccalluc** [reply]
Also, for a point of comparison, the 2020 Census did use DP, and did deliver exact enumerations at the state level, which is kind of like what you want. They did this by sacrificing consistency: summing the counts for smaller regions within a state will not match the state totals. (There's a lot more going on, and that's a simplification.)

**Reply by santidediego** [reply]
Thank you very much for the answer! it became clear to me.
